hourly-tips/src/main/java/org/apache/flink/training/exercises/hourlytips/HourlyTipsExercise.java

```diff
@@ -73,11 +89,32 @@ public JobExecutionResult execute() throws Exception {
         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
 
         // start the data generator
-        DataStream<TaxiFare> fares = env.addSource(source);
+        DataStream<TaxiFare> fares = env.addSource(source)
+                .assignTimestampsAndWatermarks(WatermarkStrategy.<TaxiFare>forMonotonousTimestamps()
+                .withTimestampAssigner(
+                        (fare, t) -> fare.getEventTimeMillis()));
+        ;
+
+        //STEP 1: Create keyed stream by driver id
+        KeyedStream<TaxiFare, Long> faresByKey = fares.keyBy((TaxiFare fare) -> fare.driverId);
+
+        //faresByKey.print();--use for testing only
+        //STEP 2: Create a window of 1 hour on that keyed stream
+        WindowedStream<TaxiFare, Long, TimeWindow> windowedKeyFares = faresByKey
+                .window(TumblingEventTimeWindows.of(Time.hours(1)));
+
+        //Invoke MaxTipsFunction on the windowed stream. This gives us all events in that 1 hour window to calculate tips
+        SingleOutputStreamOperator<Tuple3<Long, Long, Float>> totalTips = windowedKeyFares.process(new MaxTipsFunction());
+
+        //totalTips.print();//use for testing only
+
+        SingleOutputStreamOperator<Tuple3<Long, Long, Float>> hourlyMax = totalTips.windowAll(TumblingEventTimeWindows.of(Time.hours(1))).maxBy(2);
+
+        hourlyMax.addSink(sink);
 
         // replace this with your solution
         if (true) {
```

Comment:
Please remove that block, it became useless.

Reply:
Done